Sources/PackageGraph/DependencyResolver.swift

```diff
@@ -1094,3 +1153,139 @@ public class DependencyResolver<
         return container
     }
 }
+
+/// The resolver debugger.
+///
+/// Finds the constraints which results in graph being unresolvable.
+private struct ResolverDebugger<
+    Provider: PackageContainerProvider,
+    Delegate: DependencyResolverDelegate
+> where Provider.Container.Identifier == Delegate.Identifier {
+
+    typealias Identifier = Provider.Container.Identifier
+    typealias Constraint = PackageContainerConstraint<Identifier>
+
+    /// Reference to the resolver.
+    unowned let resolver: DependencyResolver<Provider, Delegate>
+
+    /// Create a new debugger.
+    init(_ resolver: DependencyResolver<Provider, Delegate>) {
+        self.resolver = resolver
+    }
+
+    // Returns the constraints which should be removed in order to make the
+    // graph resolvable.
+    func debug(
+        dependencies: [Constraint],
+        pins: [Constraint]
+    ) -> (dependencies: [Constraint], pins: [Constraint]) {
+        // Put the resolver in incomplete mode to avoid cloning new repositories.
+        resolver.isInIncompleteMode = true
+
+        let deltaAlgo = DeltaAlgorithm<ResolverChange>()
+        let allPackages = Set(dependencies.map({ $0.identifier }))
+
+        // Compute the set of changes.
+        let changes = dependencies.map({ ResolverChange.allowPackage($0.identifier) })
+                    + pins.map({ ResolverChange.allowPin($0.identifier) })
+
+        // Run the delta debugging algorithm.
+        let badChanges = deltaAlgo.run(changes: Set(changes)){ changes in
```

Comment:
Hmmm... yes, although maybe I am wrong -- maybe it is good to find the smallest set that conflicts, and then report that entire set (and where it came from, ideally).

My thinking was that it would be easier to understand "we have a bunch of the things you want, and adding any one of these other ones is going to break it, so now you just need to figure out why that is".

But it may be kind of odd that the "bunch of things you want" could include a pin or constraint you don't want, and then the thing that breaks it is the one you want...

We should probably drive this by test cases and looking at the actual diagnostics that come out, so maybe lets land as is and then consider refining as we gain experience?